aws/rust-runtime/aws-config/src/json_credentials.rs

```diff
@@ -0,0 +1,304 @@
+                })?;
+            Ok(JsonCredentials::RefreshableCredentials {
+                access_key_id,
+                secret_access_key,
+                session_token,
+                expiration,
+            })
+        }
+        Some(other) => Ok(JsonCredentials::Error {
+            code: other,
+            message: message.unwrap_or_else(|| "no message".into()),
+        }),
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::json_credentials::{
+        parse_json_credentials, InvalidJsonCredentials, JsonCredentials,
+    };
+    use std::time::{Duration, UNIX_EPOCH};
+
+    #[test]
+    fn json_credentials_success_response() {
+        let response = r#"
+        {
+          "Code" : "Success",
+          "LastUpdated" : "2021-09-17T20:57:08Z",
+          "Type" : "AWS-HMAC",
+          "AccessKeyId" : "ASIARTEST",
+          "SecretAccessKey" : "xjtest",
+          "Token" : "IQote///test",
+          "Expiration" : "2021-09-18T03:31:56Z"
+        }"#;
+        let parsed = parse_json_credentials(response).expect("valid JSON");
+        assert_eq!(
+            parsed,
+            JsonCredentials::RefreshableCredentials {
+                access_key_id: "ASIARTEST".into(),
+                secret_access_key: "xjtest".into(),
+                session_token: "IQote///test".into(),
+                expiration: UNIX_EPOCH + Duration::from_secs(1631935916),
+            }
+        )
+    }
+
+    #[test]
+    fn json_credentials_invalid_json() {
+        let error = parse_json_credentials("404: not found").expect_err("no json");
+        match error {
+            InvalidJsonCredentials::JsonError(_) => {} // ok.
+            err => panic!("incorrect error: {:?}", err),
+        }
+    }
+
+    #[test]
+    fn json_credentials_not_json_object() {
+        let error = parse_json_credentials("[1,2,3]").expect_err("no json");
+        match error {
+            InvalidJsonCredentials::JsonError(_) => {} // ok.
+            _ => panic!("incorrect error"),
+        }
+    }
+
+    #[test]
+    fn json_credentials_missing_code() {
+        let resp = r#"{
+            "LastUpdated" : "2021-09-17T20:57:08Z",
+            "Type" : "AWS-HMAC",
+            "AccessKeyId" : "ASIARTEST",
+            "SecretAccessKey" : "xjtest",
+            "Token" : "IQote///test",
+            "Expiration" : "2021-09-18T03:31:56Z"
+        }"#;
+        let parsed = parse_json_credentials(resp).expect("code not required");
+        assert_eq!(
+            parsed,
+            JsonCredentials::RefreshableCredentials {
+                access_key_id: "ASIARTEST".into(),
+                secret_access_key: "xjtest".into(),
+                session_token: "IQote///test".into(),
+                expiration: UNIX_EPOCH + Duration::from_secs(1631935916),
+            }
+        )
+    }
+
+    /*
+    #[test]
+    fn json_credentials_optional_session_token() {
```

Comment:
Why is this test commented now?

Reply:
Updated, AFAICT from the spec, token isn't actually optional